### ISSUE TITLE
Axes.fill_between() doesn't fill steps between contiguous regions

### ISSUE THREAD

**Issue by Zilwil** [matplotlib/matplotlib#9886]
### Bug report

**Bug summary**

When trying to plot a fill_between like in @https://matplotlib.org/examples/pylab_examples/fill_between_demo.html but with steps, the steps between the contiguous regions separated by the two where conditions are not filled.

**Code for reproduction**

```python
import numpy as np
from matplotlib import pyplot as plt

fig, ax = plt.subplots()

r = range(7)
a = np.array([2, 3, 3, 3, 4, 5, 6])
b = np.array([0, 0, 7, 8, 8, 2, 1])

ax.step(r, a, 'm-', where='post')
ax.step(r, b, 'b-', where='post')
ax.fill_between(r, a,b, where=a<=b, facecolor="green", step='post')
ax.fill_between(r, a,b, where=a>b, facecolor="red", step='post')

plt.show()
```

**Actual outcome**
![figure_1](https://user-images.githubusercontent.com/33915873/33411456-264fed20-d586-11e7-89f5-1b160f803146.png)

**Expected outcome**
![figure_2](https://user-images.githubusercontent.com/33915873/33411531-7e357514-d586-11e7-8bd9-245df93c214a.png)

**Proposed solution - workaround**

In Axes.py, line 4797, I changed:
```
xslice = x[ind0:ind1]
y1slice = y1[ind0:ind1]
y2slice = y2[ind0:ind1]
```
into:
```
xslice = x[ind0:ind1+1]
y1slice = y1[ind0:ind1+1]
y2slice = y2[ind0:ind1+1]
```

It solved my problem but as a matplotlib beginner I didn't check if it breaks the function behavior in some other cases.

**Matplotlib version**
  * Operating system: Debian buster 4.12.0-2
  * Matplotlib version: 2.0.0
  * Matplotlib backend: TkAgg
  * Python version: 2.7.13

Installed via apt.

**Comment by jklymak** [comment]
That’s a pretty old version.  Are you sure that’s still a bug?

**Comment by anntzer** [comment]
I can reproduce the behavior, this has nothing to do with step plots, and I believe this is the correct behavior: the first 'a>b' region ends at x=1 and the first a<=b region starts at x=2, so neither condition holds between x=1 and x=2.

**Comment by jklymak** [comment]
Thats probably whats happening, but I don't think its a very sensible way for fill-between to behave.

**Comment by anntzer** [comment]
I don't think that there `where` kwarg was a good idea to start with :/

**Comment by Zilwil** [comment]
@jklymak: The version is the Debian testing's, not even Debian stable :) I didn't test it this behavior in more recent versions, @anntzer didn't gave his version number in his answer.

@anntzer: I agree with the behavior you describe for a continuous function, and that can be smoothed with ```interpolation=True```, but not for the case of a step progression. The condition ```a<=b & a>b``` contains 7 true values, the same length as the data, maybe I'm misleading myself, but I was expecting 7 filled steps as well.
If the ```where``` arg wasn't a good start, which solution do you suggest ?

**Comment by anntzer** [comment]
That was with 2.1.
Upon further thought, I think that the *interpolation* keyword should check the value of *step* and perform, well, stepwise (rather than linear) interpolation in this case.

**Comment by Zilwil** [comment]
Yes, I tried that option thinking it would do the job. This behavior would be perfect and solve the issue.

**Comment by jklymak** [comment]
The workaround is you simply make `a[a>b]=b[a>b]` where you don't want the fill...  i.e.

```python
acopy = a.copy()
acopy[a<=b] = b[a<=b]
ax.fill_between(r, acopy, b, facecolor="green", step='post')
acopy = a.copy()
acopy[a>b] = b[a>b]
ax.fill_between(r, acopy, b, facecolor="red", step='post')
```

**Comment by jklymak** [comment]
I agree that the *where* kwarg is ambiguous, given all the other options to fill-between.  I think its better for the user to just compute the curves they want explicitly outside of the black box.

**Comment by ManuelSchneid3r** [comment]
Can confirm this bug with Python 3.6.5 matplotlib 2.2.2

**Comment by potenxyz** [comment]
This bug still exist on:
Python 3.7
matplotlib 3.2.1

The solution suggested by Will-1012 provide a good hint to solve this bug,
however, it should be modified into the following to work perfectly:

In _axes.py, line 5165:

Original code:
```
xslice = x[ind0:ind1]
y1slice = y1[ind0:ind1]
y2slice = y2[ind0:ind1]
```

Modify into:
```
xslice = x[max(ind0 - 1, 0):ind1]
y1slice = y1[max(ind0 - 1, 0):ind1]
y2slice = y2[max(ind0 - 1, 0):ind1]
```

We can test the result by modifying an example provided in [Matplotlib gallery](https://matplotlib.org/gallery/pyplots/whats_new_98_4_fill_between.html#sphx-glr-gallery-pyplots-whats-new-98-4-fill-between-py):

```
import matplotlib.pyplot as plt
import numpy as np

x = np.arange(-5, 5, 0.01)
y1 = -5*x*x + x + 10
y2 = 5*x*x + x
color_depending = np.random.rand(len(x))  # Test code

fig, ax = plt.subplots()
ax.plot(x, y1, x, y2, color='black')
# ax.fill_between(x, y1, y2, where=(y2 > y1), facecolor='yellow', alpha=0.5)  # Original code
# ax.fill_between(x, y1, y2, where=(y2 <= y1), facecolor='red', alpha=0.5)  # Original code
ax.fill_between(x, y1, y2, where=color_depending>0.5, facecolor='yellow', alpha=0.5)  # Test code
ax.fill_between(x, y1, y2, where=color_depending<=0.5, facecolor='red', alpha=0.5)  # Testcode
ax.set_title('Fill Between')

plt.show()
```

The figure looks like this, we can find that even color_depending changes fastly and random, fill_between still works well.
![Result](https://user-images.githubusercontent.com/58108403/88761224-483b3100-d1a1-11ea-9051-098964fddb9a.png)

**Comment by timhoffm** [comment]
## Clarification of the issue

If we evaluate *where*, the areas between indices with a change of the condition are neither included in `where=condition` nor in `where=not condition`.
~~~~~~~~~~~~
mask:    1 1 1 0 0 1 1 1 1
where-1: [111]     [11111]
where-0:       [0]
gap:          -   -

result:  [111]-[0]-[11111]
~~~~~~~~~~~~
This is a discretization problem, the *true* change position would be somewhere in the gap, but we don't know where exactly.

@Will-1012 solution extends to the right (with a possible overflow problem on the last index); @potenxyz extends to the left.

~~~
result:     [111]-[0]-[11111]
@Will-1012: [1111][00][11111]
@potenxyz:  [111][00][111111]
~~~

A priori, neither is universally right. For contiguous data (as in the example by @potenxyz), we have the `interpolation=True` parameter. See also https://matplotlib.org/gallery/lines_bars_and_markers/fill_between_demo.html#selectively-filling-horizontal-regions.

This does not help for step-like data as in the example by @Will-1012.

## Possible solution

I have not thought this through completely, but it seems we could/should add something like `interpolation='pre'` (and `'post' / 'mid'`) and/or `interpolation='step'` where the value is taken from the *step* parameter.

**Comment by timhoffm** [comment]
Marking as:

- new feature: The solution is a new interpolation mode.
- difficulty medium: Still needs design decisions on the API.

**Comment by QuLogic** [comment]
Since this needs some design decisions, I'm not sure it's suitable for a sprint.

**Comment by mattharrison** [comment]
Just ran into this issue today while working on my book *Effective Visualization*. I worked around the issue by adding a large overlap between the `where` parameters.
